service/storage_service.cc

```diff
@@ -1974,12 +1974,15 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
     }
 
     slogger.debug("Setting tokens to {}", bootstrap_tokens);
-    co_await mutate_token_metadata([this, &bootstrap_tokens] (mutable_token_metadata_ptr tmptr) -> future<> {
+    co_await mutate_token_metadata([this, &bootstrap_tokens, &replaced_host_id] (mutable_token_metadata_ptr tmptr) -> future<> {
         // This node must know about its chosen tokens before other nodes do
         // since they may start sending writes to this node after it gossips status = NORMAL.
         // Therefore, in case we haven't updated _token_metadata with our tokens yet, do it now.
         tmptr->update_topology(tmptr->get_my_id(), _snitch.local()->get_location(), locator::node::state::normal);
         co_await tmptr->update_normal_tokens(bootstrap_tokens, tmptr->get_my_id());
+        if (replaced_host_id) {
+            tmptr->remove_endpoint(*replaced_host_id);
```

Comment:
But this particular piece of code is only executed on the joining (replacing) node. What about other nodes in the cluster?

Reply:
Also, this seems wrong -- basically it makes `being_replaced` state not being used? But I think we do want the replaced node to stay as `being_replaced` until replace finishes.

And I don't see how it breaks the invariant you mentioned "token owners are always in normal and leaving state". And "being_replaced" **is** a leaving state, as you explained in the commit message. So the replaced node is a token owner and satisfies the invariant (it is in leaving state).

The replacing node is not yet a token owner at this point, IIUC. (It's basically bootstrapping -- tokens pending)

On the other hand, if we say that the replacing node immediately becomes the owner of these tokens (which the `update_normal_tokens` call above suggests), then indeed the invariant seems broken because of the replacing node, but I don't see how keeping or removing the replaced node from topology changes anything. We still have a normal token owner (the replacing node) that is not normal/leaving (because it's "replacing" or "bootstrapping").

Reply:
> But this particular piece of code is only executed on the joining (replacing) node. What about other nodes in the cluster?

On other nodes there is no problem. They always remove the replaced node from their topology.

> Also, this seems wrong -- basically it makes `being_replaced` state not being used? But I think we do want the replaced node to stay as `being_replaced` until replace finishes.
> 
> And I don't see how it breaks the invariant you mentioned "token owners are always in normal and leaving state". And "being_replaced" **is** a leaving state, as you explained in the commit message. So the replaced node is a token owner and satisfies the invariant (it is in leaving state).
> 
> The replacing node is not yet a token owner at this point, IIUC. (It's basically bootstrapping -- tokens pending)
> 
> On the other hand, if we say that the replacing node immediately becomes the owner of these tokens (which the `update_normal_tokens` call above suggests), then indeed the invariant seems broken because of the replacing node, but I don't see how keeping or removing the replaced node from topology changes anything. We still have a normal token owner (the replacing node) that is not normal/leaving (because it's "replacing" or "bootstrapping").

The change I did is at the end of `join_token_ring`. The replace operation has already finished, but the replacing node is still present in the topology (in the `being_replaced` state), and it will be present until the gossiper removes it (this fat client/`is_gossip_only_member` thing), which is wrong. It should be removed at the end of the joining procedure, not much later by the gossiper.

Reply:
>The replacing node is not yet a token owner at this point, IIUC. (It's basically bootstrapping -- tokens pending)

Why do you think so?

Reply:
For example, we remove the replaced node from `system.peers` even earlier:
```cpp
future<> storage_service::bootstrap(...
...
            slogger.debug("Removing replaced endpoint {} from system.peers", replace_addr);
            _sys_ks.local().remove_endpoint(replace_addr).get();
```

I chose an existing `mutate_token_metadata` call that I thought fine for simplicity. I could do it somewhere further in the joining procedure though, for example after
```cpp
// start participating in the ring.
co_await set_gossip_tokens(_gossiper, bootstrap_tokens, cdc_gen_id);
```

Reply:
> > On other nodes there is no problem. They always remove the replaced node from their topology.
>
> Is this because of node_ops RPC, which this node is not sending to itself apparently?

I don't remember. I think it happens in `storage_service::handle_state_normal`. I can check it if you want.

Reply:
`handle_state_normal` should be called on this node too (for itself). But I guess we want to fix topology faster on this node -- in the spirit of the comment above, so we don't want that state where the being_replaced node is still present (even if it's temporary).